controller/object_endpoint_controller.go

```
@@ -0,0 +1,435 @@
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+type ObjectEndpointController struct {
+	*baseController
+
+	namespace string
+	ds        *datastore.DataStore
+	image     string
+}
+
+func NewObjectEndpointController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	objectEndpointImage string,
+) *ObjectEndpointController {
+	oec := &ObjectEndpointController{
+		baseController: newBaseController("object-endpoint", logger),
+		namespace:      namespace,
+		ds:             ds,
+		image:          objectEndpointImage,
+	}
+
+	ds.ObjectEndpointInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    oec.enqueueObjectEndpoint,
+			UpdateFunc: func(old, cur interface{}) { oec.enqueueObjectEndpoint(cur) },
+			DeleteFunc: oec.enqueueObjectEndpoint,
+		},
+	)
+
+	return oec
+}
+
+func (oec *ObjectEndpointController) Run(workers int, stopCh <-chan struct{}) {
+	oec.logger.Info("Starting Longhorn Object Endpoint Controller")
+	defer oec.logger.Info("Shut down Longhorn Object Endpoint Controller")
+	defer oec.queue.ShutDown()
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(oec.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (oec *ObjectEndpointController) worker() {
+	for oec.processNextWorkItem() {
```

Check return code?

---

`for function() {}` just loops `function` so long as `function` returns something truthy.

---

Yes and it does look simple.

How is error handling modeled here? 
What does the bool return mean? End worker execution? An error that processNext.. could not handle occurred? If so, why let worker swallow that error and not bubble that up?